wandb/__init__.py

```diff
@@ -194,11 +198,32 @@ def _init_headless(api, run, job_type, cloud=True):
         stderr_redirector.redirect()
 
     # Listen on the socket waiting for the wandb process to be ready
-    while True:
-        connection, addr = server.accept()
-        res = connection.recv(128)
-        if len(res) > 0:
-            break
+    connection, addr = server.accept()
+
+    def wait(max_seconds=30):
+        started = time.time()
+        while True:
+            try:
+                res = connection.recv(128)
```

Comment:
I don't think recv is guaranteed to return all bytes, ie "ready" or "done" could potentially be split across recv. (if you look at the recv man page it says "he receive calls normally return any data available, up to
     the requested amount, rather than waiting for receipt of the full amount
     requested; this behavior is affected by the socket-level options
     SO_RCVLOWAT and SO_RCVTIMEO described in getsockopt(2)."

Reply:
Right now there are 3 messages:

1. internal_cli.py => train.py: "ready"
2. train.py => internal_cli.py: "done(exitcode)"
3. internal_cli.py => train.py: "done"

Concerning the recv issue, you think I should put it in a while loop and keep receiving if the data is less than the length of "ready"?  Maybe I should use SocketServer.TCPServer